### DISCUSSION TITLE
How do I make a 1 bit image (b/w) with a certain percentage of white ? (random noise)

### DISCUSSION THREAD

**Discussion by myicq** [ImageMagick/ImageMagick#2493] [Help] [answered]
I would like to test duty cycle on a print device, and need for this purpose a random-noise 2 bit image, with close to 80% black pixels, 20% white. It does not have to be exact. 

It's simple to create 20% gray image, but when I convert to B/W it becomes all black. 

Adding noise with "+noise random", I get around 50% white every time.

**Comment by GeeMack** [comment]
-- See edit at bottom for probably a better solution. --

ImageMagick has a range of images built in to the program that are just black and white patterns, and go in 5% steps from black to white. This command will output a 200x200 image with black and white pixels arranged to show 25% white...

    convert -size 200x200 pattern:gray25 -print "%[fx:mean]\n" result.png

In place of the "pattern:gray25" there, substitute "pattern:gray0" to "pattern:gray100" in increments of 5 to get the amount you need.
 
If the image must be random rather than a pattern, you can get that by adding "-spread" to the above command like this...

    convert -size 200x200 pattern:gray25 -virtual-pixel tile -spread 200 -print "%[fx:mean]\n" result.png

The "-print" operation there will print to the screen the mean lightness of the image so you can see the calculated percentage of white. Omit that from your final command to eliminate that output message.

See [this page about ImageMagick formats](http://www.imagemagick.org/script/formats.php) to see more of the available patterns.

-- Edited to Add -- Here is another approach. This command will make a 1000x1000 pixel image with a random distribution of black and white pixels, and within a fraction of a percent of 80% black and 20% white. Cropping any portion should result in an image with approximately the same distribution of black and white pixels. 

    convert -size 800x1000 xc:black -extent 1000x1000 -virtual-pixel tile -spread 1000 80percentblack.png

You can check the result at the command line by adding `-print "%[fx:mean]\n"` before the output file name.

**Comment by snibgo** [comment, marked as answer]
There are many methods for this. Here is another method: generate uniformly-distributed noise, and threshold it at 80%.
```
magick -size 1000x1000 xc: +noise Random -channel R -separate +channel -threshold 80% -print %[fx:mean] x.png
```

**Reply by myicq** [reply]
This command works as expected, leaving a picture with random noise with approx percentage required. Thanks!

**Comment by snibgo** [comment]
My previous solution with "-threshold" has an _expected_ proportion, that 80% of pixels will be black. The actual proportion is probably close to 80%. There is a small probability that the result is far from 80%. It could even be entirely black or entirely white, though this is highly unlikely.

Instead of using "-threshold", we can use "-contrast-stretch". IM will build a histogram, and threshold at whatever level gives the closest to 80% black. This doesn't guarantee a result of exactly 80% black, but the result is generally closer than from "-threshold".

```
magick -size 1000x1000 xc: +noise Random -channel R -separate +channel -contrast-stretch 80% -print %[fx:mean] x.png
```

**Reply by myicq** [reply]
When using contrast-stretch, there seems to be a minimum of 50%. If I use 10%, it simply defaults to 50. 
For lower range requirements, I can simply use -negate and use 90%, which is not intuitive but works.
